src/components/mask-input/mask-input.ts

```diff
@@ -0,0 +1,360 @@
+ * @fires igcFocus - Emitted when the control gains focus
+ * @fires igcBlur - Emitted when the control loses focus
+ *
+ * @csspart container - The main wrapper that holds all main input elements
+ * @csspart input - The native input element
+ * @csspart label - The native label element
+ * @csspart prefix - The prefix wrapper
+ * @csspart suffix - The suffix wrapper
+ * @csspart helper-text - The helper text wrapper
+ */
+export default class IgcMaskInputComponent extends IgcInputBaseComponent {
+  public static readonly tagName = 'igc-mask-input';
+
+  protected parser = new MaskParser();
+  protected _value = '';
+  protected selection: MaskSelection = { start: 0, end: 0 };
+  protected compositionStart = 0;
+
+  @state()
+  protected hasFocus = false;
+
+  @state()
+  protected maskedValue = '';
+
+  protected get inputSelection(): MaskSelection {
+    return {
+      start: this.input.selectionStart || 0,
+      end: this.input.selectionEnd || 0,
+    };
+  }
+
+  /** The direction attribute of the control. */
+  @property({ reflect: true })
+  public override dir: 'ltr' | 'rtl' | 'auto' = 'auto';
+
+  /** Controls the validity of the control. */
+  @property({ reflect: true, type: Boolean })
+  public invalid = false;
+
+  /**
+   * Dictates the behavior when retrieving the value of the control:
+   *
+   * - `raw` will return the clean user input.
+   * - `withFormatting` will return the value with all literals and prompts.
+   *
+   */
+  @property({ attribute: 'value-mode' })
+  public valueMode: 'raw' | 'withFormatting' = 'raw';
+
+  /**
+   * The value of the input.
+   *
+   * Regardless of the currently set `value-mode`, an empty value will return an empty string.
+   *
+   */
+  @property()
+  @blazorTwoWayBind('igcChange', 'detail')
+  public get value() {
+    return this._value
+      ? this.valueMode !== 'raw'
+        ? this.maskedValue
+        : this._value
+      : this._value;
+  }
+
+  public set value(string: string) {
+    this._value = string ?? '';
+    this.maskedValue = this.parser.apply(this._value);
+  }
+
+  /** The mask pattern to apply on the input. */
+  @property()
+  public mask!: string;
+
+  /** The prompt symbol to use for unfilled parts of the mask. */
+  @property()
+  public prompt!: string;
+
+  @watch('prompt')
+  protected promptChange() {
+    this.parser.prompt = this.prompt;
+    if (this.value) {
+      this.maskedValue = this.parser.apply(this._value);
+    }
+  }
+
+  @watch('mask')
+  protected maskChange() {
+    this.parser.mask = this.mask;
+    if (this.value) {
+      this.maskedValue = this.parser.apply(this._value);
+    }
+  }
+
+  @watch('required', { waitUntilFirstUpdate: true })
+  @watch('disabled', { waitUntilFirstUpdate: true })
+  @watch('value', { waitUntilFirstUpdate: true })
+  protected handleInvalidState() {
+    this.updateComplete.then(
+      () => (this.invalid = !this.input.checkValidity())
```

Comment:
This has the potential to override any previous invalid state from required mask positions if `required` is turned off for example as it will only check the native validity. I think you really need a single equivalent of `checkValidity` similar to the `reportValidity` for this to be consistent.

On second thought - both the existence of value and disabled should be taken into account when performing our checks for positions